### PR TITLE
Geoff/put query updates on a diet

### PR DESCRIPTION
**Pull request by gbiggs**
Closes https://github.com/open-rmf/rmf/issues/75.

### Implemented feature

Remove the query and query ID from the `MirrorUpdate` message.

### Implementation description

The query ID was previously used to know if a query update should be ignored or applied to the mirror's internal copy of the database. This is no longer necessary because query updates are separated by topic, so a mirror will only receive the updates for its registered query. Thus the query ID is removed from the `MirrorUpdate` message along with checks on it in the `MirrorManager` class.

The query was added to the query update message as part of the initial redundancy work to allow a `MirrorManager` to verify that it is receiving updates for the correct query (it hasn't subscribed to the wrong topic, or the schedule node hasn't got its queries mixed up, etc.). This is inefficient, as the query may be large and query updates are frequent. Therefore the query is removed from the query update message. To retain the functionality that it allowed, the `MirrorManager` now subscribes to the latched "registered queries" broadcast topic. Each time it receives the list of registered queries from the Schedule node, it verifies both that its own query is present and that the query ID assigned by the Schedule node matches its own opinion of the query ID.

When the query is missing, or the ID is incorrect, the `MirrorManager` will re-register its query and request a new update.

To handle possible corner cases when the Schedule node fails and the redundant backup takes over (such as a query being registered just before failure and the redundant backup not being aware of it), the `MirrorManager` will force query validation mode on when it receives the fail-over announcement. However there is a possibility of a query update from the new schedule node arriving before the fail-over announcement, which may not be for the correct query anymore. To identify this situation, the Schedule node is given an "edition" number. Each time fail-over occurs, the new schedule node gets a new edition number. The edition number is sent with each query update. The `MirrorManager` checks the received edition number, and if it is greater than the expected edition number, it also enters query validation mode. While in query validation mode, it stashes all updates. If the query is validated, then it processes the stashed updates. If the query is found to be missing or the ID is found to be incorrect, the stashed updates are dumped and a fresh update is requested.

### CODE REVIEW

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp [hunk 1]
```diff
@@ -303,6 +431,19 @@ class MirrorManager::Implementation
...
+    // 1. The next set of queries being published, and our query being in it
+    //    correctly.
+    // 2. The query update topic being subscribed to, which indicates that
+    //    our query was not stored by the schedule node correctly and we have
+    //    re-registered.
+    // While true, all query updates received will be stashed in a list, and
+    // will be popped in FIFO order and processed when this is reset to false.
+    require_query_validation = true;
+    // The new schedule node will be one edition higher
+    ++expected_node_edition;
```
Comment (mxgrey) [inline rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:445]: I don't think we should have this here since it could create a conflict with the assignment [here](https://github.com/open-rmf/rmf_ros2/blob/f6f60726f69535c8c4d010ba8980f5223e62ccd6/rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp#L275).

E.g. If we happen to process an update message from the new node with its new edition, then the `expected_node_edition` value will already be up to date. Then the fail over message may come in later and increment the value here, putting the `expected_node_edition` too high.

I think it would actually be fine to completely remove these lines here and leave it up to `update_handle` and the `queries_info_sub` callback to identify when we need to re-validate the query ID and `expected_node_edition`.

File: rmf_traffic_ros2/test/mock_monitor_nodes/missing_query_monitor.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  {
+    if (rclcpp::ok())
+    {
+      // Wait a while before announcing. This will ensure that mirrors pick up
+      // on the missing query via the query check logic, not the failover
+      // handling logic. (But the announcement still has to be made to ensure
+      // mirrors reconnect services.)
+      std::this_thread::sleep_for(5s);
+      RCLCPP_INFO(get_logger(), "Announcing fail over");
+      auto message = FailOverEvent();
```
Comment (mxgrey) [inline rmf_traffic_ros2/test/mock_monitor_nodes/missing_query_monitor.cpp:47]: I wonder if the fail over event message should include the new schedule node edition value?

**Review by mxgrey — CHANGES_REQUESTED** [review]
I haven't reviewed everything yet, but I've looked over the `MirrorManager` class pretty closely and left some feedback. This design and implementation is looking fantastic; I just have some recommendations, most of which are nitpicks, but some may be actual soundness concerns.

File: rmf_traffic_ros2/test/mock_monitor_nodes/missing_query_monitor.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  {}
+
+  void announce_fail_over()
+  {
+    if (rclcpp::ok())
+    {
+      // Wait a while before announcing. This will ensure that mirrors pick up
+      // on the missing query via the query check logic, not the failover
+      // handling logic. (But the announcement still has to be made to ensure
+      // mirrors reconnect services.)
```
Comment (mxgrey) [inline rmf_traffic_ros2/test/mock_monitor_nodes/missing_query_monitor.cpp:44]: Is this testing the potential issue that I mentioned [here](https://github.com/open-rmf/rmf_ros2/pull/86/files#r674955667)? If so, I'm surprised that we haven't seen that be an issue.

Is it possible that doing `std::this_thread::sleep_for` here is actually blocking the new schedule node from publishing its query update because this is all running on a single-threaded executor? So ultimately this `FailOverEvent` message does get published before the query update; it just all gets published after a 5s delay.

I'll try to run these test nodes and echo the topics to see what I get.

**Review by mxgrey — CHANGES_REQUESTED** [review]
I've finished reviewing the rest of the files! Just adding a few more comments. I'll start doing some manual tests of the fail over now.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp [hunk 4]
```diff
@@ -114,6 +124,63 @@ class MirrorManager::Implementation
...
+      [=](const ScheduleQueries::SharedPtr msg)
+      {
+        RCLCPP_INFO(
+          node.get_logger(),
+          "Mirror handling new sync of %d queries from schedule node",
+          msg->queries.size());
+
+        // Find what should be our query based on our query ID
+        bool found_query = false;
+        size_t ii;
```
[the remark this answers is not in the document]
Comment (mxgrey) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:141]: After looking at this again, I realize that I was wrong about `std::find_if`. But I do still recommend the change to the for-loop.

File: rmf_traffic_ros2/test/mock_monitor_nodes/missing_query_monitor.cpp (hunk 3 above)
Comment (mxgrey) [reply rmf_traffic_ros2/test/mock_monitor_nodes/missing_query_monitor.cpp:44]: Okay after testing this out manually, I understand why it's not actually triggering the edge case that I linked to above. Since the newly spawned schedule node is initially missing the query, it doesn't immediately start publishing to that query topic. Instead it publishes an empty query list, which triggers the mirror managers to re-register. However, their re-registration is held up because the `sleep_for` blocks the new schedule node from responding to the service call.

Here's where things get *wild*: Once the sleep is over, the registration requests start making it through, but the mirror manager of the monitor node destructs, which unregisters *its* query. Since it has the same query as the other mirrors that are trying to re-register, the subscriber count that we're maintaining actually drops to zero, so the new schedule node creates and then immediately discards that query, even though there are mirror managers that still want it.

At this point, query registration requests start bouncing around like crazy and the "subscription count" spikes up to ~40-60 due to massive duplication of requests from the *two* mirror managers that actually exist. The system does eventually stabilize (within about one second) and everything goes on working, but the subscription count is totally fubar and will never be able to go back down to zero.

To reproduce the issue I'm describing, I recommend running the `office.launch.xml` demo from `rmf_demos` but run the `missing_query_monitor_node` instead of the usual `rmf_traffic_schedule_monitor`. Then run these in separate terminals:
* `$ ros2 topic echo /rmf_traffic/registered_queries`
* `$ ros2 topic echo /rmf_traffic/fail_over_event`
* [`$ ros2 launch rmf_demos office_loop.launch.xml`](https://github.com/open-rmf/rmf_demos/pull/66)
* `sudo killall rmf_traffic_schedule`

You'll see some crazy output from `/rmf_traffic/registered_queries` once the message on `/rmf_traffic/fail_over_event` pops out.

I have some fixes to make the behavior a lot more sane, but I had to do a lot of debugging, so I'll need to clean that up before I push it. I should be able to do that on Monday.

An important change that we'll need is to outright get rid of the manual subscription counting. I think the whole idea is fundamentally broken in this kind of distributed system where we simply can't really know the lifecycles of the remote listeners.

An ideal alternative might be if we can give the publisher a callback that gets triggered when the number of topic subscribers changes; then we could drop the query when subscribers reach zero (and remain at zero for some time window). But it doesn't look to me like ROS2 offers that feature, so we could just take a brute force approach of doing a periodic check (every X minutes) to see if any topics have zero subscribers, and then prune them.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp [hunk 5]
```diff
@@ -114,6 +124,63 @@ class MirrorManager::Implementation
...
+  void setup_queries_sub()
+  {
+    queries_info_sub = node.create_subscription<ScheduleQueries>(
+      rmf_traffic_ros2::QueriesInfoTopicName,
+      rclcpp::SystemDefaultsQoS().reliable().keep_last(1).transient_local(),
+      [=](const ScheduleQueries::SharedPtr msg)
+      {
+        RCLCPP_INFO(
+          node.get_logger(),
+          "Mirror handling new sync of %d queries from schedule node",
```
[the remark this answers is not in the document]
Comment (mxgrey) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:136]: Addressed by #87

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mxgrey) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:141]: Addressed by #87

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp [hunk 6]
```diff
@@ -164,19 +236,53 @@ class MirrorManager::Implementation
...
+  }
+
   void handle_update(const MirrorUpdate::SharedPtr msg)
   {
     update_timer->reset();
 
-    // Validate that this patch is for the query we are expecting
-    if (convert(msg->query) != query)
+    // Verify that the expected node edition sent the update
+    if (msg->node_edition > expected_node_edition)
```
[the remark this answers is not in the document]
Comment (mxgrey) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:268]: Addressed by #87

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp [hunk 7]
```diff
@@ -164,19 +236,53 @@ class MirrorManager::Implementation
...
+      RCLCPP_WARN(
         node.get_logger(),
-        "Received update for incorrect query; re-registering query");
-      // Re-register ours to get a new topic.
-      redo_query_registration();
+        "Received query update from unexpected schedule node edition %d",
+        msg->node_edition);
+      require_query_validation = true;
+      expected_node_edition = msg->node_edition;
+    }
```
[the remark this answers is not in the document]
Comment (mxgrey) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:276]: Addressed by #87

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp (hunk 1 above)
Comment (mxgrey) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:445]: Addressed by #87

File: rmf_traffic_ros2/test/mock_monitor_nodes/missing_query_monitor.cpp (hunk 2 above)
Comment (mxgrey) [reply rmf_traffic_ros2/test/mock_monitor_nodes/missing_query_monitor.cpp:47]: Addressed by #87

File: rmf_traffic_ros2/test/mock_monitor_nodes/missing_query_monitor.cpp (hunk 3 above)
Comment (mxgrey) [reply rmf_traffic_ros2/test/mock_monitor_nodes/missing_query_monitor.cpp:44]: Addressed by #87

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MonitorNode.cpp [hunk 8]
```diff
@@ -168,6 +183,7 @@ std::shared_ptr<rclcpp::Node> MonitorNode::create_new_schedule_node()
 {
   auto database = std::make_shared<Database>(mirror.value().fork());
   auto node = std::make_shared<rmf_traffic_ros2::schedule::ScheduleNode>(
+    1, // Bump the node edition by one
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MonitorNode.cpp:186]: The monitor node, as currently designed and implemented, can only handle a single fail over. That's why I just left this at `1`.

[the remark this answers is not in the document]
Comment (mxgrey) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MonitorNode.cpp:186]: I don't see why that would be the case. If a fail over happens, we could just launch another monitor node that's set up to spawn a new schedule node with a node edition of 2. All we'd need to do is expose this parameter so that users can set it from the launch file or command line.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp [hunk 9]
```diff
@@ -164,19 +236,53 @@ class MirrorManager::Implementation
     }
   }
 
+  void process_stashed_queries()
+  {
+    RCLCPP_DEBUG(node.get_logger(), "Processing stashed queries");
+    for (auto msg: stashed_query_updates)
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp]: I have a love-hate relationship with move semantics when combined with `auto`...

Fixed (using the second option) in 34fbb74.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp [hunk 10]
```diff
@@ -70,13 +75,17 @@ class MirrorManager::Implementation
 
   rclcpp::Node& node;
   rmf_traffic::schedule::Query query;
+  uint64_t query_id = 0;
+  bool require_query_validation = false;
+  uint64_t expected_node_edition = 0;
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:80]: I initially used `version` but I felt very uncomfortable with the clash with other uses of the word. I also feel that it does not reflect what it actually is: it's not a different version of the schedule node.

Having thought more on it, would `generation` work better?

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp [hunk 11]
```diff
@@ -133,11 +200,16 @@ class MirrorManager::Implementation
...
+    require_query_validation = false;
+    process_stashed_queries();
 
-    update_timer = node.create_wall_timer(5s, [&]() -> void
-        {
-          handle_update_timeout();
-        });
+    update_timer = node.create_wall_timer(
+      5s,
+      [&]() -> void
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:209]: That just shows how easy a mistake it is to make...

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp [hunk 12]
```diff
@@ -164,19 +236,53 @@ class MirrorManager::Implementation
...
-      RCLCPP_ERROR(
+      RCLCPP_WARN(
         node.get_logger(),
-        "Received update for incorrect query; re-registering query");
-      // Re-register ours to get a new topic.
-      redo_query_registration();
+        "Received query update from unexpected schedule node edition %d",
+        msg->node_edition);
+      require_query_validation = true;
+      expected_node_edition = msg->node_edition;
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:287]: Done in 555ae53.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MonitorNode.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MonitorNode.cpp:186]: OK, I can see that argument. I don't think it's particularly feasible in practice because the launch instance that started the system wouldn't know about the node, but it is doable. I'll add a parameter.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (mxgrey) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:80]: I see where you're coming from with `generation`, but I'm concerned it's too overloaded of a word. I think in the context of programming it's more often used to refer to the "act of generating".

I don't have a super strong preference here because I think there are fair arguments to be made for any of the choices, but I do think we should be deliberate about what we choose, because the wording we go with right now could influence formal design docs and specifications in the future. What if we do a quick internal poll/discussion to see if there's any obvious preference when we take more people's views into account?

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MonitorNode.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mxgrey) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MonitorNode.cpp:186]: I can see two feasible strategies:
1. Any `MonitorNode` that gets launched after the first fail over will need to be launched manually by a human operator. The human could either input the next node edition or the new `MonitorNode` could listen for the current node edition and just increment it by 1.
2. We could pre-assign consecutive node editions to a bunch of `MonitorNode`s and launch them all at the same time. We would redesign the fail over conditions so that the `MonitorNode` will only active its schedule node if it believes that all of its predecessors have failed / timed out. This would probably be very tricky to get right, and it introduces a risk that multiple schedule nodes could get activated at the same time, which would be a disaster. So strategy (1) would probably be good for now.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:80]: A quick internal poll/discussion sounds like a good idea to me.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MonitorNode.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MonitorNode.cpp]: Done in cdaa2fd.

File: rmf_traffic_ros2/test/mock_monitor_nodes/delayed_query_broadcast_monitor.cpp [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+#include <chrono>
+
+#include <rmf_traffic_ros2/schedule/internal_MonitorNode.hpp>
+#include <rmf_traffic/schedule/Database.hpp>
+
+#include <rclcpp/rclcpp.hpp>
+
+using namespace std::chrono_literals;
+
+class DelayedQueryBroadcastScheduleNode
```
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/test/mock_monitor_nodes/delayed_query_broadcast_monitor.cpp:33]: Done in 461d732.

File: rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp:78]: The poll was inconclusive, so I will change it to `version` to be consistent.

[the remark this answers is not in the document]
Comment (gbiggs) [reply rmf_traffic_ros2/src/rmf_traffic_ros2/schedule/MirrorManager.cpp]: Done in e4b5879.
